core/src/schedule.rs

```diff
@@ -0,0 +1,180 @@
+            id_map: Default::default(),
+            id_iter: ScheduleId::MIN..,
+        }
+    }
+}
+
+impl Schedule {
+    pub fn is_empty(&self) -> bool {
+        self.dag.is_empty()
+    }
+
+    pub fn pop(&mut self) -> Option<GranId> {
+        // TODO LATER: use pop all and reorder them so modules of the same type are evaluated
+        //  after each other, that way we can keep the same wasm module alive for better performance
+
+        // schedule_id is the ID of the next node to be evaluated
+        while let Some(schedule_id) = self.dag.pop() {
+            // If this isn't in our BiBTreeMap it means that it doesn't exist anymore (it may once
+            // have been a child of an element that is now evaluated). Then we pop the next ID.
+            let Some((gran_id, _)) = self.id_map.remove_by_right(&schedule_id) else {
+                continue;
+            };
+
+            // We need to remove all children, and we do this by getting the next sibling and
+            // extracting that range. If we, for example, want to evaluate the ID 0.1, we need to
+            // remove all IDs 0.1.x, and we do that with the range 0.1..0.2. So, we find the next
+            // sibling, or if none (we are root), take the max value of the type.
+            let end_range = gran_id
+                .next_siblings()
+                .next()
+                .unwrap_or(GranularId::max_value());
+            // We collect all ScheduleIds that will now be invalid due to the pop operation
+            let removed_ids = self
+                .id_map
+                .left_range(&gran_id..&end_range)
+                .map(|(_gran, sched)| *sched)
+                .collect::<Vec<_>>();
+            // For each ID that is removed, remove it from the ID map and from all dep infos
+            removed_ids.into_iter().for_each(|id| {
+                // By removing the ID from the map, it becomes a "tombstone id" and won't be
+                // returned from further calls to this function
+                self.id_map.remove_by_right(&id);
+                // This may be optimised with other data structures
+                self.dep_info.iter_mut().for_each(|(_k, v)| {
+                    v.retain(|elem| elem.0 != id);
+                });
+            });
+            return Some(gran_id);
+        }
+
+        // In this case, we couldn't pop one element, and this may occur either if there is a
+        // cycle or if the DAG is empty. This will be checked externally.ﬁ
+        None
+    }
+
+    pub fn add_element<T, U>(
+        &mut self,
+        element: &Element,
+        ctx: &Context<T, U>,
+        format: &OutputFormat,
+    ) {
+        // Element may be one of four kinds:
+        // * Raw => We don't add it to the schedule
+        // * Module => We add it to the schedule
+        // * Parent => We add it and all its children to the schedule
+        // * Compound => We add all its children to the schedule, but not the compound itself
```

Comment:
I think that it might be nice to refactor this function in the way that the comment describes with a match expression instead of all the early returns.

Reply:
Won't be good, discussed in private with you